---
title: "Batch Effects"

output:
  html_document:
    theme: spacelab
    highlight: pygments
    code_folding: show
    toc: true
    toc_float: true
    toc_depth: 2
    df_print: paged
---

## Outline

-   Setup
-   Sources of batch effects
-   Dataset
-   Observing batch effects
-   Mitigating batch effects
-   SessionInfo

## Setup

<link rel="stylesheet" href="headers.css">

```{r, setup, include=FALSE}
# .libPaths(new = "/scratch/local/rseurat/pkg-lib-4.1.3")
.libPaths()
library(formatR)
library(knitr)
library(magrittr)
library(RColorBrewer)
library(scDblFinder)
library(SCpubr)
library(Seurat)
knitr::opts_chunk$set(fig.width = 6, fig.height = 5, message = FALSE, warning = FALSE, tidy = TRUE, tidy.opts = list(width.cutoff = 70))
```

```{r seuratVersion,warning=FALSE,message=FALSE}
library(Seurat)
packageVersion("Seurat")
```

## Sources of batch effects

Multiple steps in sample preparation introduce changes to measured gene expression.
Some of these are:

-   Cell/RNA isolation (e.g. FACS)
-   Library construction
-   Sequencing run
-   Sequencing technology
-   Individual (patient sample)
-   The person performing the experiment

## Problem description

We're going to work with a dataset comprised of pancreatic cells sequenced with 4 different single cell technologies, with various numbers of replicates.
We're first going to inspect the lowD embedding of the full dataset, and then we're going to work with a subset of the full dataset.
We're going to inspect and attempt to mitigate known batch effects in the subset dataset.


## Load the preprocessed subset dataset


Here we now use the preprocessed data from GEO, which is directly available as a Seurat Object. 


```{r load_subset}
sample <- readRDS("../../data/GSE205771/GSM6226122_oligodendroglioma_seurat_object.rds")
Seurat::DefaultAssay(sample) <- "RNA"
sample@assays$SCT <- NULL
gc()
# 

```


> âŒ¨ðŸ”¥ Exercise: Analyze this seurat object. 
- What data is saved. How many samples? how many cells? 
- What analysis layers are there?

```{r}
print("# of cells")
ncol(sample)

print("# of samples")
sample$orig.ident %>% unique %>% length()
```


We have a Analysis layer for SCT and RNA, two layers with different normalization methods

> How many cells were sequenced in each experiment?

```{r}
sample$orig.ident %>% unique   -> sample_ids
for (samp_name in sample_ids) {
  number_of_cells <- which(sample$orig.ident == samp_name) %>% length()
  paste0(samp_name, " has ", number_of_cells, " cells!") %>% print()
}
```


## Revisit UMAP

Plot UMAP embedding for the subset dataset.
Inspect the cell separation by cell type and by sequencing technology.

> âŒ¨ðŸ”¥ Exercise: plot the original UMAP by the authors with our umap. 
> Hint: group.by "orig.ident" what changes. you can choose the reduction with 
> the reduction parameter in R

```{r author umap}

do_DimPlot(sample, group.by = "orig.ident")
sample <- FindNeighbors(sample, reduction = "pca")

sample <- RunUMAP(sample, dims = 1:10, reduction = "pca", 
                  reduction.name = "umap.pca")


```

```{r unnormalized UMAP}
do_DimPlot(sample, reduction = "umap.pca", group.by = "orig.ident")
```

## What options are there to mitigate batch effects ?

Let's now explore some options of mitigating batch effects:

-   Seurat Integration
-   Seurat SCTransform
-   Conos
-   Harmony
-   ComBat/SVA
-   ...

## Seurat Integration: theory

![Seurat](../images/fx1_lrg.jpg){width="70%,height=70%"}

Seurat authors developed a strategy to "anchor" diverse datasets together: - Joint dimentional reduction - Mutual nearest neighbours ("anchors") - Filter for high-scoring anchors - Known bottleneck: rare cell types

If you want to learn more about the backround of scRNAseq integration. Have 
a look at the following review https://www.nature.com/articles/s41587-021-00895-7


## Seurat Integration: prep datasets

Prior to the integration, we want to normalize each dataset to be integrated separately.
With Seurat v5 the workflow for doing integration has changed significantly. 

```{r seurat_integrate1}

merged_seurat <- sample
# We normalize first using SCTransform and removing the effect of UMIs, genes and percent.mt.
# Process followed as in https://github.com/satijalab/seurat/issues/1679#issuecomment-557781838
merged_seurat[["RNA"]] <- split(merged_seurat[["RNA"]], f = merged_seurat$orig.ident)
merged_seurat <- NormalizeData(merged_seurat)
merged_seurat <- FindVariableFeatures(merged_seurat)
merged_seurat <- ScaleData(merged_seurat)
merged_seurat <- RunPCA(merged_seurat)
merged_seurat <- FindNeighbors(merged_seurat)
merged_seurat <- RunUMAP(merged_seurat, reduction = "pca", dims = 1:30)

# Harmony Integration. https://portals.broadinstitute.org/harmony/articles/quickstart.html
merged_seurat <- IntegrateLayers(object = merged_seurat, method = HarmonyIntegration, orig.reduction = "pca",
                       new.reduction = 'harmony', verbose = FALSE)
merged_seurat <- FindNeighbors(merged_seurat, reduction = "harmony", dims = 1:30)
merged_seurat <- FindClusters(merged_seurat, resolution = 2, cluster.name = "harmony_clusters")
merged_seurat <- RunUMAP(merged_seurat, reduction = "harmony", dims = 1:30, reduction.name = "umap.harmony")

# Anchor-based RPCA integration (method=RPCAIntegration)
merged_seurat <- IntegrateLayers(object = merged_seurat, method = RPCAIntegration,
  orig.reduction = "pca", new.reduction = 'integrated.rpca',
  verbose = FALSE)
merged_seurat <- FindNeighbors(merged_seurat, reduction = "integrated.rpca", dims = 1:30)
merged_seurat <- FindClusters(merged_seurat, graph.name = "RNA_snn",resolution = 2, cluster.name = "rpca_clusters")
merged_seurat <- RunUMAP(merged_seurat, reduction = "integrated.rpca", dims = 1:30, reduction.name = "umap.rpca")

```

The code above runs a UMAP without any integration, but also runbs integraiton 
via Harmony and Anchor based rPCA using Seurat.

Now have a look at the new low dimensional representations. 

> Exercise: Compare the different methods in the UMAP space
> How well have samples and cell types integrated? What differences do you see?

```{r, fig.width = 12}
p1 <- DimPlot(
  merged_seurat,
  reduction = "umap.rpca",
  group.by = c("orig.ident", "New_NMF_labelling", "rpca_clusters"),
  combine = FALSE, label.size = 2
)
patchwork::wrap_plots(p1)
```


## Seurat Integration


Let's clean up some redundant objects from memory.

```{r cleanup}
gc()
```



## Process the newly integrated dataset

After the integration, data scaling on the new assay is necessary, as well as calculation of PCA and UMAP embeddings.

```{r process_integrated}
# specify that we will perform downstream analysis on the corrected data note that the
# original unmodified data still resides in the 'RNA' assay

do_DimPlot(merged_seurat, reduction = "umap", group.by = "orig.ident")
do_DimPlot(merged_seurat, reduction = "umap.harmony", 
           group.by = c("orig.ident"))
do_DimPlot(merged_seurat, reduction = "umap.harmony", group.by = "New_NMF_labelling")

```

## What does the data look like after the integration?



We can also compare the distribution of cell tech labels before and after integration.

> âŒ¨ðŸ”¥ Exercise: plot gene expression after integration
>
>
> Repeat the violin and the feature plot for the dataset after integration for a gene of choice. 
> What has changed, what hasn't, and why?

The actual gene expression doesn't change, only the low dimensional graph and embeddings.

```{r}
do_FeaturePlot(merged_seurat, features = "LHFPL3", reduction = "umap.harmony")

```

The expression of this gene hasn't changed - as it is not present in the "integrated" assay, the `VlnPlot` function has fetched it's expression value from the RNA assay, which is not modified by the integration.

Let's plot it's expression on the lowD embedding:


We can now appreciate that cells highly expressing this gene are distributed over all clusters.
Note that the mixing with the cells negative for this marker is imperfect, but such is the result of the integration for this dataset.

Let's ascertain ourselves that the expression of the celltype markers is behaving as expected after integration.
Let plot 

```{r marker_after_umap,fig.width=6,fig.height=10}
# Lets take a list of marker genes from the excel file and plot some of these

excel_data <- readxl::read_excel("../1-s2.0-S2666379123004263-mmc3.xlsx", 
                                 sheet = "OD top50 markers", skip = 1) # This reads the excel file and skips the first row

# select first two genes from each column

cell_types <- colnames(excel_data)

for (cell_t in cell_types) {
  marker_genes <- excel_data[1:2,] %>% pull(cell_t)  
  do_FeaturePlot(sample, features = marker_genes, plot.title = cell_t, 
                 individual.titles = marker_genes,order = T) %>% print()
}

```

## Outlook

-   Verify that expected marker genes are expressed per cell population
-   Note: Seurat v5 has additional modalities of integrating cells: (Harmony and scVI, bridge integration across modalities)

## SessionInfo

```{r sessionInfo}
sessionInfo()
```

## Citations

BÃ¼ttner, M., Miao, Z., Wolf, F.A.
et al.
A test metric for assessing single-cell RNA-seq batch correction.
Nat Methods 16, 43--49 (2019).
<https://doi.org/10.1038/s41592-018-0254-1>

Stuart T, Butler A, Hoffman P, Hafemeister C, Papalexi E, Mauck WM 3rd, Hao Y, Stoeckius M, Smibert P, Satija R. Comprehensive Integration of Single-Cell Data.
Cell.
2019 Jun 13;177(7):1888-1902.e21.
doi: 10.1016/j.cell.2019.05.031.
Epub 2019 Jun 6.
PMID: 31178118; PMCID: PMC6687398

## Appendix


