---
title: "Batch Effects"

output:
  html_document:
    theme: spacelab
    highlight: pygments
    code_folding: show
    toc: true
    toc_float: true
    toc_depth: 2
    df_print: paged
---

## Outline

-   Setup
-   Sources of batch effects
-   Dataset
-   Observing batch effects
-   Mitigating batch effects
-   SessionInfo

## Setup

<link rel="stylesheet" href="headers.css">

```{r, setup, include=FALSE}
# .libPaths(new = "/scratch/local/rseurat/pkg-lib-4.1.3")
.libPaths()
library(formatR)
library(knitr)
library(magrittr)
library(RColorBrewer)
library(scDblFinder)
library(SCpubr)
knitr::opts_chunk$set(fig.width = 6, fig.height = 5, message = FALSE, warning = FALSE, tidy = TRUE, tidy.opts = list(width.cutoff = 70))
```

```{r seuratVersion,warning=FALSE,message=FALSE}
library(Seurat)
packageVersion("Seurat")
```

## Sources of batch effects

Multiple steps in sample preparation introduce changes to measured gene expression.
Some of these are:

-   Cell/RNA isolation (e.g. FACS)
-   Library construction
-   Sequencing run
-   Sequencing technology
-   Individual (patient sample)
-   The person performing the experiment

## Problem description

We're going to work with a dataset comprised of pancreatic cells sequenced with 4 different single cell technologies, with various numbers of replicates.
We're first going to inspect the lowD embedding of the full dataset, and then we're going to work with a subset of the full dataset.
We're going to inspect and attempt to mitigate known batch effects in the subset dataset.

## What does the data look like out of the box?

Lets have a look at the UMAP embedding of the full dataset processed in a standard way, ignoring any possible batch effects.
Compare the cell separation by sequencing technology and by cell type.

![Full dataset on umap](images/panc8full.umap.png)

## Load the preprocessed subset dataset

We have subset and pre-processed the full dataset for you.
You can download it from `zenodo`.

```{r load_subset}
sample <- readRDS("../../data/GSE205771/GSM6226122_oligodendroglioma_seurat_object.rds")
Seurat::DefaultAssay(sample) <- "RNA"

# Global parameters. Used throughout the whole analysis.
sample_name_1 <- "OE0145-IDH_NCH6341" # Name of the sample.
sample_name_2 <- "OE0145-IDH_NCH781" 

seurat_6341 <- readRDS("../sample_processed.RDS")
seurat_781 <- sample[,sample$orig.ident == sample_name_2 ]
```


```{r}
#QC

seurat_781[["percent.mt"]] <- PercentageFeatureSet(seurat_781, pattern = "^MT-")

do_ViolinPlot(
  seurat_781,
  features = c("nFeature_RNA", "nCount_RNA", "percent.mt"),
  ncol = 3, group.by = "orig.ident"
)

```

```{r featurescatter}
plot1 <- FeatureScatter(seurat_781, feature1 = "nCount_RNA", feature2 = "percent.mt", group.by = "orig.ident")
plot2 <- FeatureScatter(seurat_781, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", group.by = "orig.ident")
plot3 <- FeatureScatter(seurat_781, feature1 = "nCount_RNA", feature2 = "SPARCL1",slot = "counts", group.by = "orig.ident")
plot1 / plot2
plot3
```

```{r filter, echo=FALSE}

# Perform QC on the sample.
# Get cutoffs.
counts_lower_cutoff <- 1000 # Minimum amount of UMIs per cell.
genes_lower_cutoff <- 500 # Minimum amount of genes per cell.
mito_higher_cutoff <- 5 # Maximum amount of mitochondrial RNA per cell.

# Generate the first subset.
count_mask <- seurat_781$nCount_RNA > counts_lower_cutoff
gene_mask <- seurat_781$nFeature_RNA > genes_lower_cutoff
mito_mask <- seurat_781$percent.mt < mito_higher_cutoff
mask <- count_mask & gene_mask & mito_mask
summary(mask)
seurat_781 <- seurat_781[, mask] # Subset the sample.

# Upper cutoffs are determined by the mean and standard deviation of the remaining cells.
counts_higher_cutoff <- mean(seurat_781$nCount_RNA) + 3 * stats::sd(seurat_781$nCount_RNA)
genes_higher_cutoff <- mean(seurat_781$nFeature_RNA) + 3 * stats::sd(seurat_781$nFeature_RNA)


# Second subset based on the mean and standard deviation of the remaining cells.
count_mask <- seurat_781$nCount_RNA < counts_higher_cutoff
gene_mask <- seurat_781$nFeature_RNA < genes_higher_cutoff
mask <- count_mask & gene_mask
summary(mask)

seurat_781 <- seurat_781[,mask]
```


```{r}
sce = scDblFinder(
    SingleCellExperiment(
        list(counts=seurat_781@assays$RNA$counts),
    ) 
)
doublet_score = sce$scDblFinder.score
doublet_class = sce$scDblFinder.class
seurat_781$scDblFinder.score <- sce$scDblFinder.score
seurat_781$scDblFinder.class <- sce$scDblFinder.class
hist(seurat_781$scDblFinder.score, breaks = "FD")
FeatureScatter(seurat_781, feature1 = "nCount_RNA", feature2 = "nFeature_RNA",group.by = "scDblFinder.class")
```
> ðŸ§­âœ¨ Poll:
>
> [Which sequencing technologies are retained in the subset dataset?](https://PollEv.com/multiple_choice_polls/j2zBwetamS5edfxq0JDs8/respond)
>
> How many cells were sequenced in each experiment?

## Revisit UMAP

Plot UMAP embedding for the subset dataset.
Inspect the cell separation by cell type and by sequencing technology.

```{r umap_sub, echo=FALSE,fig.width=12,fig.show='hold',fig.align='center'}

```

> âŒ¨ðŸ”¥ Exercise: plot gene expression
>
> In the previous course units, you have learned to call differentially expressed genes with Seurat.
> In this task, we ask you to: - call genes differentially expressed between cells sequenced with the smartseq2 technology and those with the celseq technology - plot a violin plot for the top gene - plot a feature plot for the top gene

> ðŸ§­âœ¨ Poll: [What is the name of the top DE gene ?](https://PollEv.com/multiple_choice_polls/D5BfavArbiXgs8QeHvDNn/respond)

Here's our proposed solution:

```{r merge normalize and plot}
merged_seurat <- merge(seurat_6341, seurat_781)

merged_seurat <- NormalizeData(merged_seurat)
merged_seurat <- FindVariableFeatures(merged_seurat, selection.method = "vst", nfeatures = 2000) 

(plot1 <- VariableFeaturePlot(merged_seurat))

(plot2 <-
  LabelPoints(
    plot = plot1,
    points = head(VariableFeatures(merged_seurat), 10),
    repel = TRUE
  ))
merged_seurat <- ScaleData(merged_seurat, features = rownames(merged_seurat))


```
```{r runpca}
merged_seurat <- RunPCA(merged_seurat, features = VariableFeatures(object = merged_seurat))

do_DimPlot(merged_seurat, reduction = "pca", group.by = "orig.ident",plot.axes = T)
ElbowPlot(merged_seurat)
merged_seurat <- FindNeighbors(merged_seurat, dims = seq_len(8))
merged_seurat <- RunUMAP(merged_seurat, dims = 1:8, n.neighbors = 20, verbose = FALSE,)
do_DimPlot(merged_seurat, reduction = "umap", group.by = "orig.ident")
```

## What options are there to mitigate batch effects ?

Let's now explore some options of mitigating batch effects:

-   Seurat Integration
-   Seurat SCTransform
-   Conos
-   Harmony
-   ComBat/SVA
-   ...

## Seurat Integration: theory

![Seurat](images/fx1_lrg.jpg){width="70%,height=70%"}

Seurat authors developed a strategy to "anchor" diverse datasets together: - Joint dimentional reduction - Mutual nearest neighbours ("anchors") - Filter for high-scoring anchors - Known bottleneck: rare cell types

## Seurat Integration: prep datasets

Prior to the integration, we want to normalize each dataset to be integrated separately.

```{r seurat_integrate1}
npcs <- 10
# Normalize using Harmony. The effect to remove is the orig.ident metadata column (from where each sample comes from).
sample <- harmony::RunHarmony(sample, dims = 1:npcs, group.by.vars = "orig.ident", assay.use = "RNA")

```

## Seurat Integration

In this paragraph, we're going to run the three functions key to Seurat Integration.

```{r seurat_integrate2}
# select features that are repeatedly variable across datasets for integration
features <- SelectIntegrationFeatures(object.list = panc.list)

anchors <- FindIntegrationAnchors(object.list = panc.list, anchor.features = features, normalization.method = "LogNormalize")

# this command creates an 'integrated' data assay
panc.combined <- IntegrateData(anchorset = anchors, normalization.method = "LogNormalize")
```

Let's clean up some redundant objects from memory.

```{r cleanup}
rm(list = c("panc.list", "features", "anchors"))
gc()
```

Let's have a brief look at the panc.combined dataset - a new Assay has been created by the integration procedure.

> ðŸ§­âœ¨ Poll: What is the new assay called?]
> (<https://PollEv.com/multiple_choice_polls/4ArWqQjwtyVafBYCiaqbr/respond>) Hint: you can access assays of a Seurat object with `Assays()`.

## Process the newly integrated dataset

After the integration, data scaling on the new assay is necessary, as well as calculation of PCA and UMAP embeddings.

```{r process_integrated}
# specify that we will perform downstream analysis on the corrected data note that the
# original unmodified data still resides in the 'RNA' assay
DefaultAssay(panc.combined) <- "integrated"

# Run the standard workflow for visualization and clustering
panc.combined <- ScaleData(panc.combined, verbose = FALSE)
panc.combined <- RunPCA(panc.combined, npcs = 30, verbose = FALSE)
panc.combined <- RunUMAP(panc.combined, reduction = "pca", dims = 1:30)
```

## What does the data look like after the integration?

Let's have a look at the distribution of technology, and celltype, labels in lowD after integration.

```{r integrated_umap, echo=FALSE,fig.width=12,fig.show='hold',fig.align='center'}
p1 <- DimPlot(panc.combined, reduction = "umap")
p2 <- DimPlot(panc.combined, reduction = "umap", group.by = "celltype")
p1 + p2
```

We can also compare the distribution of cell tech labels before and after integration.

```{r b_a_umap, echo=FALSE,fig.width=12,fig.show='hold',fig.align='center'}
p1 <- DimPlot(panc_sub, reduction = "umap")
p2 <- DimPlot(panc.combined, reduction = "umap")
p1 + p2
```

> âŒ¨ðŸ”¥ Exercise: plot gene expression after integration
>
> You have previously plotted the expression of a DE gene between the sequencing technologies on the dataset before integration.
> Repeat the violin and the feature plot for the dataset after integration.
> What has changed, what hasn't, and why?

Here's our proposed solution:

```{r solution_DEG_after}
VlnPlot(panc.combined, features = "LRRC75A-AS1", group.by = "celltype", split.by = "tech")
```

The expression of this gene hasn't changed - as it is not present in the "integrated" assay, the `VlnPlot` function has fetched it's expression value from the RNA assay, which is not modified by the integration.

Let's plot it's expression on the lowD embedding:

```{r DEG_after_umap,fig.width=6,fig.height=10}
p1 <- FeaturePlot(panc.combined, features = "LRRC75A-AS1")
p2 <- DimPlot(panc.combined, reduction = "umap", group.by = "tech", cols = scales::alpha(my_cols, 0.3))
p1 + p2
```

We can now appreciate that cells highly expressing this gene are distributed over all clusters.
Note that the mixing with the cells negative for this marker is imperfect, but such is the result of the integration for this dataset.

Let's ascertain ourselves that the expression of the celltype markers is behaving as expected after integration.
For this purpose, let's plot the expression of the alpha cell marker gen "GCG".

```{r marker_after_umap,fig.width=6,fig.height=10}
p1 <- FeaturePlot(panc.combined, features = "GCG")
p2 <- DimPlot(panc.combined, reduction = "umap", group.by = "celltype")
p1 + p2
```

## Outlook

-   Verify that expected marker genes are expressed per cell population
-   Note: Seurat v5 has additional modalities of integrating cells: (Harmony and scVI, bridge integration across modalities)

## SessionInfo

```{r sessionInfo}
sessionInfo()
```

## Citations

BÃ¼ttner, M., Miao, Z., Wolf, F.A.
et al.
A test metric for assessing single-cell RNA-seq batch correction.
Nat Methods 16, 43--49 (2019).
<https://doi.org/10.1038/s41592-018-0254-1>

Stuart T, Butler A, Hoffman P, Hafemeister C, Papalexi E, Mauck WM 3rd, Hao Y, Stoeckius M, Smibert P, Satija R. Comprehensive Integration of Single-Cell Data.
Cell.
2019 Jun 13;177(7):1888-1902.e21.
doi: 10.1016/j.cell.2019.05.031.
Epub 2019 Jun 6.
PMID: 31178118; PMCID: PMC6687398

## Appendix

This is how to reproduce generation of the UMAP embedding for the full dataset and further the generation of the subset dataset used in this course unit:

```{=html}
<code>
library(SeuratData)
panc <- LoadData("panc8", type = "default")
table(panc$dataset)
panc <- NormalizeData(panc)
panc <- FindVariableFeatures(panc, selection.method = "vst", nfeatures = 2000)
all.genes <- rownames(panc)
panc <- ScaleData(panc, features = all.genes)
panc <- RunPCA(panc, features = VariableFeatures(object = panc))
panc <- RunUMAP(panc, dims = 1:15)
p1<-DimPlot(panc, reduction = "umap",label=TRUE)
p2<-DimPlot(panc, reduction = "umap",group.by="celltype")
p1+p2

panc_sub <- subset(x=panc,subset = dataset == c("celseq","celseq2","smartseq2"))
panc_sub <- NormalizeData(panc_sub)
panc_sub <- FindVariableFeatures(panc_sub, selection.method = "vst", nfeatures = 2000)
all.genes <- rownames(panc_sub)
panc_sub <- ScaleData(panc_sub, features = all.genes)
panc_sub <- RunPCA(panc_sub, features = VariableFeatures(object = panc_sub))
panc_sub <- RunUMAP(panc_sub, dims = 1:15)
saveRDS(panc_sub,"panc_sub_processed.RDS")
</code>
```
