---
title: "First Steps into a Regular Analysis"

output:
  html_document:
    theme: spacelab
    highlight: pygments
    code_folding: show
    toc: true
    toc_float: true
    toc_depth: 2
    df_print: paged
---

# Setup

Welcome to the first steps of the RNAseq Course. 

Each document covers part of the analysis pipeline. We will be using the Seurat 
package for the analysis.
```{r setup}

# This setup block loads the required packages for the course
# These should be preinstalled
suppressMessages({
  library(tidyverse)
  library(Seurat)
  library(tidyverse)
  library(scDblFinder)
  library(SCpubr)
})

set.seed(8211673)

knitr::opts_chunk$set(echo = TRUE, format = TRUE, out.width = "100%")

# Here we setup the multithreading setup for R 
options(
  parallelly.fork.enable = FALSE,
  future.globals.maxSize = 8 * 1024^2 * 1000
)

plan("multicore", workers = 8) # function made available by SeuratObj automatically.
```

The commands in this chunk show the current WD which in Rstudio and Rmarkdown is the position of the document. 

```{r useful-information, echo=FALSE}
cat("work directory: ", getwd())
cat("\n")
cat("library path(s): ", .libPaths())
```

## Links in case you need help

- Seurat documentation: [Seurat](https://satijalab.org/seurat/)
- SCpubR documentation: [SCpubR](https://enblacar.github.io/SCpubr-book/)
- Rstudio Cheatsheets: [Rstudio Cheatsheets](https://www.rstudio.com/resources/cheatsheets/)

# Load Data

We will be analyzing a dataset from the following paper: Blanco-Carmona, E. et al. Tumor heterogeneity and tumor-microglia interactions in primary and recurrent IDH1-mutant gliomas. Cell Reports Medicine 4, 101249 (2023).https://www.sciencedirect.com/science/article/pii/S2666379123004263

We will focus on a few oligodendroglioma which were sequenced here in Heidelberg, however the processed 
data is available in GEO: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE205771

The primary analysis was run by the DKFZ Omics Data Core Facility (ODCF) which is also
helping provide the infrastructure for this course.

For further details on the primary analysis pipeline that gives you the count data, please head over to [cellranger website](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger).

After the cellranger pipeline has been row, one way to load the data is using these files
-   `matrix.mtx`: **count matrix** represents the number of molecules for each gene (row) that are detected in each cell (column).
-   `genes.tsv`: a list of ENSEMBL-IDs and their corresponding gene symbol
-   `barcodes.tsv`: a list of molecular barcodes that identifies each cell uniquely
(the alternative is loading the h5 encoded data. Either way can be done in Seurat)
https://satijalab.org/seurat/reference/read10x
https://satijalab.org/seurat/reference/read10x_h5 


We are working on machines provided by the DENBI cloud. The main data directory should be shown
ob the beamer (or just ask where on the server the data is stored. )

For now we assume that this data resides in a directory `datasets/filtered_gene_bc_matrices/hg19` (relative to this current markdown file).


Alternatively, R/Bioconductor provides tools to download data directly from GEO. Here
we have ane example on how to download data from our Publication directly to the hard drive. 

```{r download from geo}
options(timeout = max(300, getOption("timeout")))
options(download.file.method.GEOquery = "wget")
#GEOquery::getGEOSuppFiles("GSE205771",baseDir = "/omics/groups/OE0436/internal/heyer/scourse/sc_course/data/",)

```

#  Quality control {.unnumbered}

This section, titled "Quality Control", is part of an R Markdown document. 
 The count matrix from cellranger is located using sample_path. 
 Instead of creating a new Seurat object from the count matrix, a pre-existing Seurat object is read from an RDS file. 
 The Seurat object is subset to include only cells from the sample of interest 
 The percentage of mitochondrial RNA in the cells, a common quality control metric, is computed. 
 Finally, a heatmap is generated to visualize the first few genes and cells.

```{r}
#| eval: true
#| 
# Enrique Blanco Carmona
# e.blancocarmona@kitz-heidelberg.de
# PhD Student â€“ Clinical Bioinformatics
# Division of Pediatric Neurooncology (B062)
# DKFZ-KiTZ | Germany

#--------------------------------------------------------------------
# 1 - READ IN COUNT MATRICES AND INDIVIDUAL QC
#--------------------------------------------------------------------

# ------------------------------------------------------------------
# All the following code including doublet detection has to be run
# for each sample individually.
# ------------------------------------------------------------------

# Global parameters. Used throughout the whole analysis.
sample_name <- "OE0145-IDH_NCH6341" # Name of the sample.

# Load samples.
mt_pattern <- "^MT-"
sample_path <- "/vol/data/course_data/data/filtered_feature_bc_matrix" # Path to where the count matrix from cellranger (either filtered or raw, your choice) is located.


# Read the count matrix and generate a Seurat object.
sample <- Seurat::Read10X(data.dir = sample_path)  # 
sample <- Seurat::CreateSeuratObject(sample, project = sample_name, min.cells = 3, min.features = 200)




```


This data is extremely big and sparse, this variable is now an object of type `dgCMatrix`.
Sparse Matrices save a lot of memory by only storing the non-zero values.
Consider; if we saved a matrix of 10000x10000 with only 1% of the values being non-zero, we would save 99% of the memory by using a sparse matrix.

Lets examine a few genes in the first thirty cells:


And, we can have a heatmap of the first few genes and cells. For 
this we use the "pheatmap" function from the pheatmap package. 

```{r explore2, echo=FALSE, out.width="50%"}
pheatmap::pheatmap(log10(sample@assays$RNA$counts[1:500, 1:200] + 1),
  show_rownames = F,
  show_colnames = F,
  cluster_rows = F,
  cluster_cols = F,
  color = colorRampPalette(
    c("lightgrey", "red")
  )(100)
) # note: colorRampPalette returns another function.
```

> ðŸ§­âœ¨ Task:
>
> How many genes and cells does this dataset have?
>
> How many genes are not expressed in any cell?
> Hint: subset genes with `rowSums()`.
>
> Which are the top 3 genes with the highest total count?
> Hint: `rowSums()` and `sort`.

> âŒ¨ðŸ”¥ Exercise: Plot the histogram of counts for the previous top gene over all cells

```{r acces counts}
sample@assays$RNA@layers$counts 

```


# Seurat Object
>  **Note:** usually on any data science context, we refer to our columns as features.
> This is not the case for the count matrix.

The `min.cells` and `min.features` arguments are first low-stringency **filters**.

The `SeuratObject` serves as a container that contains both data (like the count matrix) and analysis (like PCA, or clustering results) for a single-cell dataset.
For example, the count matrix is stored in `sample[["RNA"]]@counts`.
On RStudio, you can use `View(sample)` to inspect all the slots.

At the top level, `SeuratObject` serves as a collection of `Assay` and `DimReduc` objects, representing expression data and dimensional reductions of the expression data, respectively.
The `Assay` objects are designed to hold expression data of a single type, such as RNA-seq gene expression, CITE-seq ADTs, cell hashtags, or imputed gene values.

On the other hand, `DimReduc` objects represent transformations of the data contained within the Assay object(s) via various dimensional reduction techniques such as PCA.
For class-specific details, including more in depth description of the slots, please see the wiki sections for each class:

-   [`Seurat`](https://github.com/satijalab/seurat/wiki/Seurat)

    -   [Slots](https://github.com/satijalab/seurat/wiki/Seurat#slots)
    -   [Object Information](https://github.com/satijalab/seurat/wiki/Seurat#object-information)
    -   [Data Access](https://github.com/satijalab/seurat/wiki/Seurat#data-access)

-   [`Assay`](https://satijalab.org/seurat/articles/announcements.html)

    -   [Slots](https://satijalab.org/seurat/articles/announcements.html)
    -   [Object Information](https://github.com/satijalab/seurat/wiki/Assay#object-information)
    -   [Data Access](https://github.com/satijalab/seurat/wiki/Assay#data-access)

-   [`DimReduc`](https://github.com/satijalab/seurat/wiki/DimReduc)

    -   [Slots](https://github.com/satijalab/seurat/wiki/DimReduc#slots)
    -   [Object Information](https://github.com/satijalab/seurat/wiki/DimReduc#object-information)
    -   [Data Access](https://github.com/satijalab/seurat/wiki/DimReduc#data-access)


When accessing data from a Seurat object we use the `@` operator.
For example with `sample@assays$RNA$counts` we access the count matrix.
We can view metadata with `sample@meta.data`.

# Quality Control

One of our first goals is to identify (and filter) dead cells that could be the results of a harsh experimental protocol.
A few QC metrics commonly used, include:

1.  **The number of unique genes detected in each cell.**

-   Low-quality cells or empty droplets will often have very few genes.
-   Cell doublets or multiplets may exhibit an aberrant high gene count.

1.  Similarly, **the total number of molecules detected within a cell (correlates strongly with unique genes)**
2.  **The percentage of reads that map to the mitochondrial genome.**

-   Low-quality / dying cells often exhibit extensive mitochondrial contamination.
-   We use the set of all genes starting with MT- as a set of mitochondrial genes.

For further details, see [this publication](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4758103/).

The number of unique genes and total molecules are automatically calculated during `CreateSeuratObject()`.
You can find them stored in the object `meta.data`, let's see for the first 5 cells:

```{r metadata}
sample@meta.data %>% head(5)
```

The `@` operator we just used, is for accessing the slot on the object.

The `[[` operator can add columns to object metadata.
This is a great place to stash additional QC stats:

```{r getMT}
#library(Seurat)
sample[["percent.mt"]] <- PercentageFeatureSet(sample, pattern = "^MT-")
```

`PercentageFeatureSet()` function calculates the percentage of counts originating from a set of features.
In the example above we can easily access all miochondrial genes because their names start with "\^MT".
So we give this as pattern (aka *regular expression*).

Let's visualize the distribution of these metrics over all cells (as Violin plots):

```{r vlnplot}
# ScPubr plotting function
do_ViolinPlot(
  sample,
  features = c("nFeature_RNA", "nCount_RNA", "percent.mt"),
  ncol = 3, group.by = "orig.ident"
)
# Standard Seurat plotting function
VlnPlot(
  sample,
  features = c("nFeature_RNA", "nCount_RNA", "percent.mt"),
  ncol = 3
)
```

The `VlnPlot()` function plots the probability density function for all the specified variables (features).
Note the similarities to SCpubR. SCpubR provides some nice defaults and many options in changing the code. 


> ðŸ§­âœ¨ Task: How many cells have less than 2000 counts in total (summed over all genes)?
> Hint: use `colSums`.

Individually these variables may not fully discriminate dead cells, but could also reflect real biological properties (e.g. higher mitochondrial count).
Therefore it is useful to look a relationship between these variables.
`FeatureScatter()` is typically used to visualize relationships between features, but it can also be used for anything calculated at the object, i.e. columns in object metadata or for genes (rows in the count matrix).
All those are **features**

```{r featurescatter}
plot1 <- FeatureScatter(sample, feature1 = "nCount_RNA", feature2 = "percent.mt", group.by = "orig.ident")
plot2 <- FeatureScatter(sample, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", group.by = "orig.ident")
plot3 <- FeatureScatter(sample, feature1 = "nCount_RNA", feature2 = "SPARCL1",slot = "counts", group.by = "orig.ident")

# Here we can use functionality from the "patchwork" package to combine plots with / or \ operators
plot1 / plot2
plot3
```

> Exercise: 

# Filtering and Transformation

## Select Cells

Based on cell-specific features we can subset our `SeuratObject` to keep only the 'cells' in good state.
In this case, based on the previous Violin plots, we'll use the following criteria:

> Excericise:
> Based on the plots above -> Choose your cutoffs

```{r filter, echo=FALSE}

# Perform QC on the sample.
# Get cutoffs.
counts_lower_cutoff <- 1000 # Minimum amount of UMIs per cell.
genes_lower_cutoff <- 500 # Minimum amount of genes per cell.
mito_higher_cutoff <- 0.05 # Maximum amount of mitochondrial RNA per cell.

# Generate the first subset.
count_mask <- sample$nCount_RNA > counts_lower_cutoff
gene_mask <- sample$nFeature_RNA > genes_lower_cutoff
mito_mask <- sample$percent.mt < mito_higher_cutoff
mask <- count_mask & gene_mask & mito_mask
summary(mask)
sample <- sample[, mask] # Subset the sample.

# Upper cutoffs are determined by the mean and standard deviation of the remaining cells.
counts_higher_cutoff <- mean(sample$nCount_RNA) + 3 * stats::sd(sample$nCount_RNA)
genes_higher_cutoff <- mean(sample$nFeature_RNA) + 3 * stats::sd(sample$nFeature_RNA)


# Second subset based on the mean and standard deviation of the remaining cells.
count_mask <- sample$nCount_RNA < counts_higher_cutoff
gene_mask <- sample$nFeature_RNA < genes_higher_cutoff
mask <- count_mask & gene_mask
summary(mask)

sample <- sample[,mask]
```

> ðŸ§­âœ¨ Task: What's the current number of cells after this step?

## Doublet Finding

Checking for doublets.

```{r}

sce = scDblFinder(
    SingleCellExperiment(
        list(counts=sample@assays$RNA$counts),
    ) 
)
doublet_score = sce$scDblFinder.score
doublet_class = sce$scDblFinder.class
sample$scDblFinder.score <- sce$scDblFinder.score
sample$scDblFinder.class <- sce$scDblFinder.class
hist(sample$scDblFinder.score, breaks = "FD")
FeatureScatter(sample, feature1 = "nCount_RNA", feature2 = "nFeature_RNA",group.by = "scDblFinder.class")
```


## Normalization

After removing unwanted cells from the dataset, the next step is to normalize the data.
By default, we employ a global-scaling normalization method "LogNormalize" that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10000 by default), and log-transforms the result.
Normalized values are stored in `sample[["RNA"]]@data`.

```{r norm}
sample <- NormalizeData(sample,)
```

> Excercise: Now plot the distribution of the normalized data across the dataset. 

## Informative Genes

The main goal is to select genes that will help us to organize cells according to the transcription profile, this are the genes that will be in the spotlight for our following step.
Therefore we look for a subset of genes ("features") that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others).

To identify the most highly variable genes, Seurat models the mean-variance relationship inherent in the data using the `FindVariableFeatures()` function.
By default, it uses the `vst` methodology with 2000 features per dataset.

First, fits a line to the relationship of `log(variance)` and `log(mean)` using local polynomial regression (`loess`).
Then standardizes the feature values using the observed mean and expected variance (given by the fitted line).
Feature variance is then calculated on the standardized values after clipping to a maximum (by default, square root of the number of cells).
These will be used downstream in dimensional reductions like PCA.

<!--Some relevant publications for Feature selection and its importance are: [[1](https://www.nature.com/articles/nmeth.2645)], and [[2](https://doi.org/10.1016/j.cell.2019.05.031)].-->

```{r HVG}
sample <- FindVariableFeatures(sample, selection.method = "vst", 
                                nfeatures = 2000) 
```

<!--Note that there's also the `selection.method = "mvp"` that identifies variable features while controlling for the strong relationship between variability and average expression.-->

> ðŸ§­âœ¨ Task:
>
> Which are the 3 most highly variable genes?
> Hint: use `VariableFeatures()`.
>
> What's the variance of the gene `PYCARD`?
> hint: use `HVFInfo()`.

Plot variable features:

```{r HVG_plot}
(plot1 <- VariableFeaturePlot(sample))
```

Now with labels, taking top10 genes as in the recent question:

```{r HVG_LabelPlot}
(plot2 <-
  LabelPoints(
    plot = plot1,
    points = head(VariableFeatures(sample), 10),
    repel = TRUE
  ))
```

## Scaling

Next, we apply a linear transformation ('scaling') that is a standard pre-processing step prior to dimensional reduction techniques like PCA.
The `ScaleData()` function:

-   Shifts the expression of each gene, so that the mean expression across cells is `0`
-   Scales the expression of each gene, so that the variance across cells is `1`. This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate.
-   more generally one can also model the mean expression as a function of other variables from the metadata, i.e. *regress them out* before scaling the residuals (see: `vars.to.regress`)
-   The results of this are stored in `sample[["RNA"]]@scale.data`

```{r scaledata}
sample <- ScaleData(sample, features = rownames(sample))
```

# Dimensional Reduction

Next we perform PCA on the scaled data.
By default, only the previously determined variable features are used as input, but can be defined using features argument if you wish to choose a different subset.

```{r runpca}
sample <- RunPCA(sample, features = VariableFeatures(object = sample))
```

> Do you feel like you need a refresher on PCA?
> check [StatQuest with Josh Starmer video](https://youtu.be/FgakZw6K1QQ) explaining PCA by SVD step by step!
> (duration: 20 minutes)

Examine and visualize PCA results a few different ways:

```{r dimplot}
DimPlot(sample, reduction = "pca",group.by = "scDblFinder.class") + NoLegend()

# Alternative SCPubr Plotting function
do_DimPlot(sample, reduction = "pca", group.by = "scDblFinder.class",plot.axes = T)
```

```{r printpca}
print(sample[["pca"]], dims = 1:5, nfeatures = 5)
```

```{r vizdimloadings}
VizDimLoadings(sample, dims = 1:2, reduction = "pca")
```

In particular `DimHeatmap()` allows for easy exploration of the primary sources of heterogeneity in a dataset, and can be useful when trying to decide which PCs to include for further downstream analyses.
Both cells and features are ordered according to their PCA scores.
Setting cells to a number plots the 'extreme' cells on both ends of the spectrum, which dramatically speeds plotting for large datasets.
Though clearly a supervised analysis, we find this to be a valuable tool for exploring correlated feature sets.

```{r dimheatmap}
DimHeatmap(sample, dims = 1:9, cells = 500, balanced = TRUE)
```

To overcome the extensive technical noise in any single gene for scRNA-seq data, Seurat clusters cells based on their PCA scores.
Here each PC essentially represents a 'metagene' that combines information across a correlated gene sets.
The top principal components therefore represent a robust compression of the dataset.

One quick way to determine the 'dimensionality' of the dataset is by eyeballing how the percentage of variance explained decreases:

```{r elbowplot}
ElbowPlot(sample)
```

> ðŸ§­âœ¨ Task: How many components should we choose to include?

When picking the 'elbow' point, remember that **it's better to err on the higher side**!
Also, if your research questions aim towards rare celltypes, you may definitely include more PCs (think about it in terms of the variance in gene expression values).

<!-- Another methodology is using the JackStraw procedure: We randomly permute a subset of the data (1% by default) and rerun PCA, constructing a 'null distribution' of feature scores, and repeat this procedure. We identify 'significant' PCs as those who have a strong enrichment of low p-value features. Of course, such permutations tend to be cumbersome and this is a highly intense computational procedure, so we're skipping it in the course. -->

# Playing around with metadata

Let's have a look again into the values we got in `percent.mt`:

```{r}
hist(sample$percent.mt)
```

```{r}
summary(sample$percent.mt)
```

Now, let's add a column annotating samples with low, medium or high `percent.mt`

```{r}
sample$mt.categories <- NA

sample$mt.categories[sample$percent.mt <= 1] <- "Low"
sample$mt.categories[sample$percent.mt > 1 &
                     sample$percent.mt <= 3 ] <- "Medium"
sample$mt.categories[sample$percent.mt > 3] <- "High"

stopifnot(all(! is.na(sample$percent.mt)))
```

Let's explore what we just did:

```{r mt_categories}
VlnPlot(sample,
  features = "percent.mt",
  group.by = "mt.categories",
  sort = "decreasing"
) +
  ggtitle(NULL) + NoLegend()

# Alternative with SCpubR
do_ViolinPlot(sample, features = "percent.mt", group.by = "mt.categories",)
```

Finally, we are able to plot PCA, and have shape by these categories:

```{r}
DimPlot(sample, shape.by = "mt.categories")
do_DimPlot(sample,reduction = "pca", group.by = "mt.categories")
```

What about using colors?

To do that, we need to 'speak' to the Seurat function (`DimPlot`) through the `SeuratObj`.
There is a variable (actually, it's a slot) inside it, called `active.ident`.
To access or set this to different values (e.g. metadata columns) we use the `Idents()` function:

```{r}
Idents(sample) %>% head(3)
```

By default, all our cells have one single value, that was set while creating the Seurat object in our early steps.

Let's set our new metadata column as the new active identity, and let `DimPlot` play out it's straightforward behavior:

```{r}
Idents(sample) <- "mt.categories"

DimPlot(sample)
```

> Exercise: Choose another metadata column and analyze if it may be influencing our count data. 

# Cell Cycle Regression 

# Cell Cycle Regression 

In single-cell RNA-seq analysis, cell cycle regression is a crucial step to remove 
unwanted variation attributed to cell cycle effects. In this section, we use the 
Seurat package to perform cell cycle regression. We first identify S phase and 
G2-M phase genes using Seurat's updated 2019 cell cycle gene sets. We then score 
each cell in our dataset based on its expression of these genes using the 
CellCycleScoring function. This assigns each cell an S phase score and a G2-M 
phase score. Next, we perform Principal Component Analysis (PCA) using these cell 
cycle genes. Finally, we regress out the cell cycle effects using the ScaleData 
function in Seurat, which helps to ensure that downstream analysis is not 
confounded by cell cycle effects. It's important to note that the active identity 
of the Seurat object should be tracked throughout this process.

```{r}
# Get S phase genes.
s.genes <- Seurat::cc.genes.updated.2019$s.genes
# Get G2-M phase genes.
g2m.genes <- Seurat::cc.genes.updated.2019$g2m.genes

sample <- CellCycleScoring(sample, s.features = s.genes, g2m.features = g2m.genes, 
set.ident = TRUE)

#RidgePlot(sample, features =, ncol = 2,assay = "RNA",layer = "counts")
```

```{r}
sample <- RunPCA(sample, features = c(s.genes, g2m.genes))
DimPlot(sample)
```

Whether cell cycle regression is actually necessary depends on the dataset and its structure. 
Always inspect your data first before you decide whether or not cell cycle regression (or the regression of other metadata such as nCount_RNA ) is necessary.

```{r}
sample <- ScaleData(sample, vars.to.regress = c("S.Score", "G2M.Score"), features = rownames(sample))
sample <- RunPCA(sample, features = VariableFeatures(object = sample))

DimPlot(sample, reduction = "pca")

```

MAKE SURE TO SAVE your sample so we can use it later tomorrow. 

```{r}
# Finally save the sample object using the saveRDS function to ensure that we can load it later in the other documents
saveRDS(object = sample, "../sample_processed.RDS")
```

# End

```{r sessionInfo}
sessionInfo()
```
